src/suggestion.c

```diff
@@ -0,0 +1,448 @@
+// and append both to the prefix
+int try_swap(match_t **set, trie_t * t, char *prefix, char *suffix, int edits_left, int n) {
+
+    int rc = EXIT_SUCCESS;
+    int len = strlen(prefix);
+    char* new_prefix;
+
+    if (len == 0 || strlen(suffix) == 0) {
+        return EXIT_SUCCESS;
+    }
+
+    new_prefix = malloc(sizeof(char) * (MAXLEN + 1));
+    if (new_prefix == NULL) {
+        return EXIT_FAILURE;
+    }
+
+    strncpy(new_prefix, prefix, MAXLEN);
+
+    // Swap suffix's first character into new_prefix and insert it in front of the last character
+    new_prefix[len] = new_prefix[len - 1];
+    new_prefix[len - 1] = suffix[0];
+    new_prefix[len + 1] = '\0';
+
+    if (has_children(t, new_prefix) == EXIT_SUCCESS) {
+        // Adding 1 to the suffix pointer will essentially delete the first character
+        rc = suggestions(set, t, new_prefix, suffix + 1, edits_left - 1, n);
+    }
+
+    // Save some space now that we're done
+    free(new_prefix);
+
+    if (rc != EXIT_SUCCESS) {
+        return EXIT_FAILURE;
+    }
+
+    return EXIT_SUCCESS;
+}
+
+// Helper function for suggestions that tries to insert a character to the end of a prefix
+int try_insert(match_t **set, trie_t *t, char *prefix, char *suffix, int edits_left, int n) {
+
+    int i;
+    int rc = EXIT_SUCCESS;
+    int len = strlen(prefix);
+    char* new_prefix;
+
+    // Ran out of space
+    if (len == MAXLEN - 1) {
+        return EXIT_FAILURE;
+    }
+
+    for (i = 1; i < 248; i++) {
+
+        char c = (char)i;
+
+        if (trie_char_exists(t, c) == EXIT_SUCCESS) {
+            
+            new_prefix = malloc(sizeof(char) * (MAXLEN + 1));
+            if (new_prefix == NULL) {
+
+                return EXIT_FAILURE;
+            
+            } else {
+
+                strncpy(new_prefix, prefix, MAXLEN);
+
+                // Try adding on a new character
+                new_prefix[len] = c;
+                new_prefix[len + 1] = '\0';
+
+                if (has_children(t, new_prefix) == EXIT_SUCCESS) {
+
+                    // Basically just inserting the new ASCII character to the string
+                    rc = suggestions(set, t, new_prefix, suffix, edits_left - 1, n);
+
+                    if (rc != EXIT_SUCCESS) {
+
+                        free(new_prefix);
+                        return EXIT_FAILURE;
+                    }
+                } 
+            
+                // Save some space now that we're done
+                free(new_prefix);
+            }
+        }
+    }
+
+    // EXIT_FAILURE would've been caught above, so if we make it here we succeeded
+    return EXIT_SUCCESS;
+}
+
+// Helper function for suggestions(). Attempts to add a match to a suggestion set
+int attempt_add(match_t **set, trie_t *t, char *s, int edits_left, int n) {
+    int i;
+
+    // Check if the current string is in the trie
+    if (trie_search(t, s) == 1) {
```

Comment:
Let's compare against the constants `IN_TRIE`, `PARTIAL_IN_TRIE`, and `NOT_IN_TRIE`.

Reply:
Updated to check against IN_TRIE